libpod/container_inspect.go

```diff
@@ -493,3 +925,492 @@ func (c *Container) generateInspectContainerConfig(spec *spec.Spec) (*InspectCon
 
 	return ctrConfig, nil
 }
+
+// Generate the InspectContainerHostConfig struct for the HostConfig field of
+// Inspect.
+func (c *Container) generateInspectContainerHostConfig(ctrSpec *spec.Spec, namedVolumes []*ContainerNamedVolume, mounts []spec.Mount) (*InspectContainerHostConfig, error) {
+	hostConfig := new(InspectContainerHostConfig)
+
+	logConfig := new(InspectLogConfig)
+	logConfig.Type = c.config.LogDriver
+	hostConfig.LogConfig = logConfig
+
+	restartPolicy := new(InspectRestartPolicy)
+	restartPolicy.Name = c.config.RestartPolicy
+	restartPolicy.MaximumRetryCount = c.config.RestartRetries
+	hostConfig.RestartPolicy = restartPolicy
+
+	hostConfig.Dns = make([]string, 0, len(c.config.DNSServer))
+	for _, dns := range c.config.DNSServer {
+		hostConfig.Dns = append(hostConfig.Dns, dns.String())
+	}
+
+	hostConfig.DnsOptions = make([]string, 0, len(c.config.DNSOption))
+	for _, opt := range c.config.DNSOption {
+		hostConfig.DnsOptions = append(hostConfig.DnsOptions, opt)
+	}
+
+	hostConfig.DnsSearch = make([]string, 0, len(c.config.DNSSearch))
+	for _, search := range c.config.DNSSearch {
+		hostConfig.DnsSearch = append(hostConfig.DnsSearch, search)
+	}
+
+	hostConfig.ExtraHosts = make([]string, 0, len(c.config.HostAdd))
+	for _, host := range c.config.HostAdd {
+		hostConfig.ExtraHosts = append(hostConfig.ExtraHosts, host)
+	}
+
+	hostConfig.GroupAdd = make([]string, 0, len(c.config.Groups))
+	for _, group := range c.config.Groups {
+		hostConfig.GroupAdd = append(hostConfig.GroupAdd, group)
+	}
+
+	if ctrSpec.Process != nil {
+		if ctrSpec.Process.OOMScoreAdj != nil {
+			hostConfig.OomScoreAdj = *ctrSpec.Process.OOMScoreAdj
+		}
+	}
+
+	hostConfig.ReadonlyRootfs = ctrSpec.Root.Readonly
+	hostConfig.ShmSize = c.config.ShmSize
+	hostConfig.Runtime = "oci"
+
+	// This is very expensive to initialize.
+	// So we don't want to initialize it unless we absolutely have to - IE,
+	// there are things that require a major:minor to path translation.
+	var deviceNodes map[string]string
```

Comment:
Nit of a nit: [sync.Once](https://golang.org/pkg/sync/#Once) can help hiding such details in the implementation.

Reply:
That actually breaks error propagation - `Do` can't return errors that occur during the function. We can wire up manual error handling with closures, but at that point we're at roughly the same line count.